Formula/v8.rb

```diff
@@ -15,98 +16,213 @@ class V8 < Formula
-    (buildpath/"third_party/icu").install resource("icu")
-    (buildpath/"buildtools").install resource("buildtools")
-    (buildpath/"base/trace_event/common").install resource("common")
-    (buildpath/"tools/swarming_client").install resource("swarming_client")
-    (buildpath/"testing/gtest").install resource("gtest")
-    (buildpath/"testing/gmock").install resource("gmock")
-    (buildpath/"tools/clang").install resource("clang")
-
-    system "make", "native", "library=shared", "snapshot=on",
-                   "console=readline", "i18nsupport=off",
-                   "strictaliasing=off"
-
-    include.install Dir["include/*"]
-
-    cd "out/native" do
-      rm ["libgmock.a", "libgtest.a"]
-      lib.install Dir["lib*"]
-      bin.install "d8", "mksnapshot", "process", "shell" => "v8"
+    output_path = "out.gn/x64.release"
+
+    # Trick gclient to reuse our v8 checkouts
+    base = "gclient-cache/chromium.googlesource.com"
+    (buildpath/"#{base}-v8-v8").install Dir["*"]
+
+    (resources - [resource("depot_tools")]).each do |r|
+      (buildpath/"#{base}-#{r.name}").install r
+    end
+
+    # Install depot_tools in PATH
+    (buildpath/"depot_tools").install resource("depot_tools")
+    ENV.prepend_path "PATH", buildpath/"depot_tools"
+
+    # Prevent from updating depot_tools on every call
+    # see https://www.chromium.org/developers/how-tos/depottools#TOC-Disabling-auto-update
+    ENV["DEPOT_TOOLS_UPDATE"] = "0"
+
+    # Initialize and sync gclient
+    system "gclient", "root"
+    system "gclient", "config", "--spec", <<~EOS
+      solutions = [
+        {
+          "url": "https://chromium.googlesource.com/v8/v8.git",
+          "managed": False,
+          "name": "v8",
+          "deps_file": "DEPS",
+          "custom_deps": {},
+        },
+      ]
+      target_os = [ "mac" ]
+      target_os_only = True
+    EOS
+
+    system "gclient", "sync",
+      "--cache-dir=#{buildpath}/gclient-cache",
+      "-j", ENV.make_jobs,
+      "-r", version,
+      "--no-history",
+      "-vvv"
+
+    # Enter the v8 checkout
+    cd "v8" do
+      gn_args = {
+        :is_debug => false,
+        :is_component_build => true,
+        :v8_use_external_startup_data => false,
+        :v8_enable_i18n_support => true,
+      }
+
+      # Transform to args string
+      gn_args_string = gn_args.map { |k, v| "#{k}=#{v}" }.join(" ")
+
+      # Build with gn + ninja
+      system "gn",
+        "gen",
+        "--args=#{gn_args_string}",
+        output_path
+
+      system "ninja",
+        "-j", ENV.make_jobs,
+        "-C", output_path,
+        "-v",
+        "d8"
+
+      # Install all the things
+      include.install Dir["include/*"]
+      cd output_path do
+        lib.install Dir["lib*.dylib"]
+
+        # Install d8 and icudtl.dat in libexec and symlink
+        # because they need to be in the same directory.
+        libexec.install Dir["d8", "icudt*.dat"]
+        mkdir_p bin
+        (bin/"d8").atomic_write <<~EOS
```

Comment:
Why `atomic_write` instead of `.write`? There's no chance of `bin/d8` already existing, no?

Reply:
Not sure exactly. That was copy/pasted from a comment from @chrmoritz earlier in the PR. https://github.com/Homebrew/homebrew-core/pull/25786#discussion_r179010862

Reply:
Let's try a `.write`, not least because then the `mkdir_p bin` can go 👋.